Function that can load a nexus event file using scipp/mantid and

SimonHeybrock opened the pull request:
expand this to more pixels / events by replicating the existing
pixels / events. Random noise is added to the event times in order
to avoid perfect overlap in that regard.

A quick test script is available that just runs the function with
a reasonable set of inputs.

No data file is included as all data files we have available can
not be shared due to the ISIS data policy. A datafile for inhouse
testing is available through ownCloud.

mads-bertelsen committed:
Function that can load a nexus event file using scipp/mantid and
expand this to more pixels / events by replicating the existing
pixels / events. Random noise is added to the event times in order
to avoid perfect overlap in that regard.

A quick test script is available that just runs the function with
a reasonable set of inputs.

No data file is included as all data files we have available can
not be shared due to the ISIS data policy. A datafile for inhouse
testing is available through ownCloud.

SimonHeybrock commented on expand_load.py:
```diff
@@ -0,0 +1,135 @@
+        original_id = 0
+        for pixel_id in range(n_pixels):
+            # For each pixel, grab the events from the original file
+            pos_n_events = len(np.array(d["loaded_data"].coords[Dim.Tof].values[original_id]))
```
I recommend working with `DataArray` instead of `Dataset`. Then you can avoid the artificial (and more lengthy) explicit item names such as `"loaded_data"`.

Why did you convert to `np.array` before taking `len`?

mads-bertelsen replied:
Will try to use DataArray in the future, but I don't see how I could have done it in that particular instance? Should I have saved  d["loaded_data"] as another variable and worked with that?

I didn't know scipp had len functionality built in, it is never mentioned in the documentation.

SimonHeybrock replied:
Something like:
```python
loaded_data = sc.neutron.load(...)
output = sc.DataArray(coords=...)
```

mads-bertelsen replied:
I see, thought you meant internally, but the returned result can just be a DataArray. This branch has been updated with this improvement now. Ready to merge?

SimonHeybrock commented on test_expand_load.py:
```diff
@@ -0,0 +1,18 @@
+#data = expand_load.expand_data_file('../../wish_event_data/WISH00043525.nxs', 6E5)
+#data = expand_load.expand_data_file('../../wish_event_data/WISH00043525.nxs', 3E5, n_events=6E8, verbose=True)
+#data = expand_load.expand_data_file('../../wish_event_data/WISH00043525.nxs', 3E5, n_events=6E8, time_noise_us=500, verbose=True)
+#data = expand_load.expand_data_file('../../wish_event_data/WISH00043525.nxs', 3E5, n_events=6E8, time_noise_us=100, verbose=True)
```
I suggest we remove the hardcoded filenames and use a command-line argument as the filename to load, so we can include this in `master`.

mads-bertelsen replied:
It has been updated with a command line interface including all options and documentation with both doctoring and -h option.

SimonHeybrock commented on expand_load.py:
```diff
@@ -0,0 +1,135 @@
+
+            # If any events present, add random noise and append them as sparse data
+            if pos_n_events > 0:
+                noise = np.random.normal(pos_n_events)*time_noise_us
```
Was there a reason for not using the numpy parameter `scale=time_noise_us`?

mads-bertelsen replied:
No reason, just didn't know that option existed. Will use that.

SimonHeybrock reviewed (commented):
@mads-bertelsen Some small comments below.

Additionally, the main thing I think we need to fix is that we lose extra information from the loaded input that may be essential. For example the `component_info` labels are lost, which contain source and sample positions. That is, we could not convert units of the created dataset.

To fix this, I think we should try to be future-proof (so you do not need to come back and "fix" this function 5 times): In the coming weeks/months we will probably extend the existing converters and add other information that is present in the Mantid workspace. The main example that comes to mind is information about the sample and other beamline logs. That is, we need to avoid implementing `expand_data_file` in a "white list" approach (copying things explicitly by hand). I think one way to do this is to just create a copy of the original `Dataset` (or `DataArray`), `del` the things that we know depend on the spectrum (TOFs, coords[`Dim.Position`], and labels[`spectrum_number`]), and add them back in (as you do now).

mads-bertelsen committed:
Changed the test script to take command line arguments. This is
documented in the file, as a DocString and with the -h argument.

Included component_info from loaded data in the final dataset.

Avoided unnecessary use of numpy when finding length of scipp variable.

Updated to use of scale keyword argument in numpy.

mads-bertelsen commented:
All concerns have been addressed, except the change to a "blacklist" approach over the current "whitelist" approach, which was not possible due to scipp not being able to delete coords / labels. The loaded component_info was however forwarded to the new dataset.

SimonHeybrock commented:
I have created https://github.com/scipp/scipp/issues/564 for the missing functionality.

mads-bertelsen committed:
Changed from use of Dataset to DataArray as the output is simpler.
